File: ios/RecipeClipper/Data/Backup/Backup.swift
import Foundation

// One export file: every recipe, list and membership, as plain data. Ported from Android's
// data/backup/Backup.kt; both platforms test against the same fixture files in
// shared/fixtures/backup/, which is what makes an export from one readable on the other.
//
// Every record carries a stable `id`: the row's `uid`, which never changes on the phone that
// made it (a rename or a re-share keeps it). Memberships refer to those ids, never to database
// row ids.
//
// Deliberately not in the file: cached photos (only `imageUrl`) and cook progress (timers, the
// current step, the chosen servings), which live in memory today and, if #10 persists them,
// are still a moment in one kitchen rather than part of the recipe.

struct Backup: Equatable {
    /// The `format` marker: tells an export apart from any other JSON file.
    static let format = "recipe-clipper-backup"

    /// Bumped only when an older app would misread a newer file. Adding a field or a whole new
    /// top-level section (a meal plan, say) doesn't need a bump: readers ignore keys they don't
    /// know, and import only ever adds.
    static let formatVersion = 1

    var exportedAt: Int64
    var recipes: [BackupRecipe]
    var lists: [BackupList]
    var memberships: [BackupMembership]
    /// The pantry (#51). Absent in older files, which read as empty.
    var pantry: [BackupPantryItem] = []
    /// The grocery list (#50), in list order. Absent in older files, which read as empty.
    var groceries: [BackupGroceryItem] = []
}

struct BackupRecipe: Equatable {
    var id: String
    var sourceUrl: String
    /// "BLOG" or "REDDIT"; anything else is read as BLOG when imported.
    var sourceType: String
    var title: String
    var imageUrl: String?
    var ingredients: [String]
    var instructions: [String]
    var prepTime: String?
    var cookTime: String?
    var totalTime: String?
    var servings: String?
    var lastViewedAt: Int64
    var checkedIngredients: Set<Int>
    var notes: String?
    /// The recipe's language tag (#14), so an import reads it with the same words.
    var language: String? = nil
    /// Whose words the content is (#29): "PARSED" (absent in older files), "EDITED", "CLIPPED"
    /// or "MANUAL". Kept so an imported edit is still never refreshed by a re-share.
    var contentOrigin: String = "PARSED"
    /// When the user last saved an edit, or nil.
    var editedAt: Int64? = nil
}

struct BackupList: Equatable {
    var id: String
    var name: String
    var isFavorites: Bool
    var isBuiltIn: Bool
    var sortOrder: Int
    var createdAt: Int64
}

struct BackupMembership: Equatable {
    var recipeId: String
    var listId: String
    var addedAt: Int64
}

/// A pantry item (#51). The days are epoch days; `aisle` an `Aisle` key.
struct BackupPantryItem: Equatable {
    var id: String
    var name: String
    var quantity: String?
    var language: String?
    var aisle: String
    var inStock: Bool
    var alwaysHave: Bool
    var purchasedDay: Int64?
    var expiresDay: Int64?
    var updatedAt: Int64
}

/// A grocery item (#50). `recipeId` is the file id of the recipe it came from, or nil (typed,
/// or its recipe is gone); `plannedDay` the planned day it came from, if any.
struct BackupGroceryItem: Equatable {
    var id: String
    var text: String
    var language: String?
    var aisle: String
    var checked: Bool
    var recipeId: String?
    var plannedDay: Int64?
    var updatedAt: Int64
}

/// Why an export or an import failed, as a cause: the Settings screen picks the words. An import
/// that fails for any of these has written nothing.
enum BackupError: Error, Equatable {
    /// Not JSON, not an object, no `format` marker, or far too big to be an export.
    case notABackup
    /// Made by a newer app whose format this one can't read.
    case newerVersion(found: Int)
    /// An export, but damaged: the detail names the first bad field (diagnostic, not copy).
    case malformed(String)
    /// The picked file couldn't be opened or read.
    case readFailed
    /// The database refused the import. It ran in one transaction, so nothing was written.
    case saveFailed
    /// The recipes couldn't be read out, or the file couldn't be written.
    case exportFailed
}

/// What an import did, for the one-line summary.
struct ImportSummary: Equatable {
    /// New recipes written.
    var recipesAdded: Int
    /// New lists created (lists that combined with one already here don't count).
    var listsAdded: Int
    /// Recipes in the file that were already on this phone (same cleaned link).
    var recipesAlreadyHere: Int
    /// Recipes in no list left out because history was full (see BackupMerger).
    var recipesSkipped: Int
    /// New pantry items written (#51).
    var pantryAdded = 0
    /// New grocery items written (#50).
    var groceriesAdded = 0
}

/// An export ready to hand to the share sheet.
struct ExportedBackup: Equatable {
    var json: String
    var exportedAt: Int64
    var recipeCount: Int
}

File: app/src/main/java/com/example/recipeclipper/data/backup/Backup.kt
package com.example.recipeclipper.data.backup

/**
 * One export file: every recipe, every list and every membership, as plain data. The JSON
 * shape is in [BackupJson]; the rules for bringing one into a phone that already has recipes
 * are in [BackupMerger]. Pure types, no Android: the iOS app has the same ones
 * (`Data/Backup/Backup.swift`) and both platforms test against the same fixture files in
 * `shared/fixtures/backup/`.
 *
 * Every record carries a stable id ([BackupRecipe.id], [BackupList.id], [BackupPantryItem.id],
 * [BackupGroceryItem.id]): the row's `uid`, which
 * never changes on the phone that made it (a rename or a re-share keeps it). Memberships refer
 * to those ids, never to database row ids.
 *
 * Deliberately not in the file: cached photos (only [BackupRecipe.imageUrl]) and cook progress
 * (timers, the current step, the chosen servings), which live in memory today and, if #10
 * persists them, are still a moment in one kitchen rather than part of the recipe.
 */
data class Backup(
    val exportedAt: Long,
    val recipes: List<BackupRecipe>,
    val lists: List<BackupList>,
    val memberships: List<BackupMembership>,
    /** The pantry (#51). Absent in older files, which read as empty. */
    val pantry: List<BackupPantryItem> = emptyList(),
    /** The grocery list (#50), in list order. Absent in older files, which read as empty. */
    val groceries: List<BackupGroceryItem> = emptyList()
) {
    companion object {
        /** The `format` marker: tells an export apart from any other JSON file. */
        const val FORMAT = "recipe-clipper-backup"

        /**
         * Bumped only when an older app would misread a newer file. Adding a field or a whole
         * new top-level section (a meal plan, say) doesn't need a bump: readers ignore keys
         * they don't know, and import only ever adds.
         */
        const val FORMAT_VERSION = 1
    }
}

data class BackupRecipe(
    val id: String,
    val sourceUrl: String,
    /** "BLOG" or "REDDIT"; anything else is read as BLOG when imported. */
    val sourceType: String,
    val title: String,
    val imageUrl: String?,
    val ingredients: List<String>,
    val instructions: List<String>,
    val prepTime: String?,
    val cookTime: String?,
    val totalTime: String?,
    val servings: String?,
    val lastViewedAt: Long,
    val checkedIngredients: Set<Int>,
    val notes: String?,
    /** The recipe's language tag (#14), so an import reads it with the same words. */
    val language: String? = null,
    /** Whose words the content is (#29): "PARSED" (absent in older files), "EDITED",
     *  "CLIPPED" or "MANUAL". Kept so an imported edit is still never refreshed by a re-share. */
    val contentOrigin: String = "PARSED",
    /** When the user last saved an edit, or null. */
    val editedAt: Long? = null
)

data class BackupList(
    val id: String,
    val name: String,
    val isFavorites: Boolean,
    val isBuiltIn: Boolean,
    val sortOrder: Int,
    val createdAt: Long
)

data class BackupMembership(
    val recipeId: String,
    val listId: String,
    val addedAt: Long
)

/** A pantry item (#51). [purchasedDay] and [expiresDay] are epoch days; [aisle] an `Aisle` key. */
data class BackupPantryItem(
    val id: String,
    val name: String,
    val quantity: String?,
    val language: String?,
    val aisle: String,
    val inStock: Boolean,
    val alwaysHave: Boolean,
    val purchasedDay: Long?,
    val expiresDay: Long?,
    val updatedAt: Long
)

/**
 * A grocery item (#50). [recipeId] is the file id of the recipe it came from, or null (typed,
 * or its recipe is gone); [plannedDay] the planned day (an epoch day) it came from, if any.
 */
data class BackupGroceryItem(
    val id: String,
    val text: String,
    val language: String?,
    val aisle: String,
    val checked: Boolean,
    val recipeId: String?,
    val plannedDay: Long?,
    val updatedAt: Long
)

/**
 * Why an export or an import failed, as a cause: the Settings screen picks the words. An
 * import that fails for any of these has written nothing.
 */
sealed class BackupError {
    /** Not JSON, not an object, no `format` marker, or far too big to be an export. */
    object NotABackup : BackupError()

    /** Made by a newer app whose format this one can't read. */
    data class NewerVersion(val found: Int) : BackupError()

    /** An export, but damaged: [detail] names the first bad field (diagnostic, not copy). */
    data class Malformed(val detail: String) : BackupError()

    /** The picked file couldn't be opened or read. */
    object ReadFailed : BackupError()

    /** The database refused the import. It ran in one transaction, so nothing was written. */
    object SaveFailed : BackupError()

    /** The recipes couldn't be read out, or the file couldn't be written. */
    object ExportFailed : BackupError()
}

sealed class BackupResult<out T> {
    data class Success<T>(val value: T) : BackupResult<T>()
    data class Failure(val error: BackupError) : BackupResult<Nothing>()
}

/** What an import did, for the one-line summary. */
data class ImportSummary(
    /** New recipes written. */
    val recipesAdded: Int,
    /** New lists created (lists that combined with one already here don't count). */
    val listsAdded: Int,
    /** Recipes in the file that were already on this phone (same cleaned link). */
    val recipesAlreadyHere: Int,
    /** Recipes in no list left out because history was full (see [BackupMerger]). */
    val recipesSkipped: Int,
    /** New pantry items written (#51). */
    val pantryAdded: Int = 0,
    /** New grocery items written (#50). */
    val groceriesAdded: Int = 0
)

/** An export ready to hand to the share sheet. */
data class ExportedBackup(val json: String, val exportedAt: Long, val recipeCount: Int)
